packages/state-history/src/lib/entities-state-history.spec.ts

```diff
@@ -0,0 +1,171 @@
+import { stateHistory } from './state-history';
+import { Store, createStore, withProps } from '@ngneat/elf';
+import {
+  withEntities,
+  withUIEntities,
+  UIEntitiesRef,
+  entitiesPropsFactory,
+  DefaultEntitiesRef,
+  EntitiesState,
+  setEntities,
+  getAllEntities,
+  updateEntities,
+  deleteEntities,
+  addEntities,
+} from '@ngneat/elf-entities';
+import {
+  entitiesStateHistory,
+  EntitiesStateHistory,
+} from './entities-state-history';
+import { expectTypeOf } from 'expect-type';
+
+function eq(store: Store, value: number) {
+  expect(store.getValue()).toEqual({ prop: value });
+}
+
+interface Entity {
+  id: number;
+  label: string;
+}
+
+describe('entities state history', () => {
+  it('should be able to be used only with entities', () => {
+    const propsStore = createStore(
+      { name: '' },
+      withProps<{ loading?: boolean }>({})
+    );
+    // It doesn't work without entities
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-expect-error
+    entitiesStateHistory(propsStore);
+
+    const entitiesStore = createStore({ name: '' }, withEntities<Entity>());
+    const entitiesHistory = entitiesStateHistory(entitiesStore);
+
+    expectTypeOf(entitiesHistory).toEqualTypeOf<
+      EntitiesStateHistory<typeof entitiesStore, DefaultEntitiesRef>
+    >();
+
+    const uiEntitiesStore = createStore(
+      { name: '' },
+      withUIEntities<Entity, 'label'>({ idKey: 'label' })
+    );
+    const uiEntitiesHistory = entitiesStateHistory(uiEntitiesStore, {
+      entitiesRef: UIEntitiesRef,
+    });
+
+    expectTypeOf(uiEntitiesHistory).toEqualTypeOf<
+      EntitiesStateHistory<typeof uiEntitiesStore, typeof UIEntitiesRef>
+    >();
+
+    const { withProductEntities, productEntitiesRef } =
+      entitiesPropsFactory('product');
+
+    const productEntitiesStore = createStore(
+      { name: '' },
+      withProductEntities<{ id: string }>()
+    );
+    const productEntitiesHistory = entitiesStateHistory(productEntitiesStore, {
+      entitiesRef: productEntitiesRef,
+    });
+
+    expectTypeOf(productEntitiesHistory).toEqualTypeOf<
+      EntitiesStateHistory<
+        typeof productEntitiesStore,
+        typeof productEntitiesRef
+      >
+    >();
+  });
+
+  it('should set the current state', () => {
+    const store = createStore({ name: '' }, withEntities<Entity>());
+    const history = entitiesStateHistory(store);
+
+    expect((history as any).entitiesHistoryMap).toEqual({});
+  });
+
+  it('should work', () => {
+    const store = createStore({ name: '' }, withEntities<Entity>());
+    const history = entitiesStateHistory(store);
+
+    store.update(
+      setEntities([
+        { id: 1, label: 'first' },
+        { id: 2, label: 'second' },
+      ])
+    );
+    history.undo(1);
+
```

Comment:
I'm not sure why this is the result? You call `undo(1)`, so the next line should be `toEqual([])`, no?

Reply:
It performs undo for the entity with id = 1. As it was just added it hasn't any history so `history.undo(1)` does nothing. It is one of the questions. Should I delete it if I evoke `undo` on just created entity?

Reply:
Oh, so the methods are working on entity id?

Reply:
yes, like Akita`s one but with different internals

Reply:
Now I remember :)

I think it should behave like Akita - https://opensource.salesforce.com/akita/docs/plugins/state-history#api-1

Reply:
> Should I delete it if I evoke undo on just created entity?

I don't remember the behavior in Akita. I'd just go with the exact same behavior as Akita.

Reply:
I must check out what the Akita behave for just created entity

Reply:
Sure. You can play/run the specs and see the result. Note that also in Akita, the ids are optional.

Reply:
It is! I'll go deeper into it and come back with changes

Reply:
As I understand, Akita does not perform undoing for deleting and adding, it works only for updating. But I feel like it's a good feature, especially if you do optimistic updates on deleting or creating an entity.

Reply:
Are you sure? what happens if we add an entity and call `undo()` in Akita?

Reply:
I used playground(updated akita) and just added a plugin https://stackblitz.com/edit/akita-todos-app-se7grz?file=src%2Fapp%2Ftodos%2Fstate%2Ftodos.service.ts

Reply:
I used example from site and updated Akita
https://stackblitz.com/edit/akita-todos-app-se7grz?file=src%2Fapp%2Ftodos%2Fstate%2Ftodos.service.ts,src%2Fapp%2Ftodos%2Fstate%2Ftodos.query.ts

Reply:
Cool. If you think you can support it, it'll be great.

Reply:
So leave it with adding and deleting? And what about just a created entity, can we undo it?

Reply:
- Add entity/ Set entity => undo() => remove it
- Delete entity => undo() => re-add it

Reply:
and support optional ids

Reply:
I've done it but I don't like it much. There's a problem with bringing items back to the array. They are placed in the end even if they were deleted when they were somewhere in the middle of all entities. Sorting is not an option because people may use server-side pagination-soring(as I do 😄). I guess Akita could have declined to implement this feature because of the same problem. I believe it can be handled by plain `stateHistory`. I don't see any other way to do it simply and workable, sorry

Reply:
You can complete the specs and docs.

Reply:
I'm refactoring logic a little bit and when I'm done I'll doo the others stuff

Reply:
It was more than `a little`😄. I made it more similar to Akita. Added `clearPast` and `clearFuture` if you want to clear `past` but do not want to make 1 more update to set `past` after clearing. Now it's even more typed. If you're ok with this one and tests as well I will start adding docs.
But please take time to play with it, maybe there are some bugs(😄 I hope not)